### ISSUE TITLE
Update for Bootstrap v4

### ISSUE THREAD

**Issue by cvrebert** [twbs/bootlint#304]

**Comment by zacechola** [comment]
:+1:

**Comment by kkirsche** [comment]
:+1: when do you think is the right time to do the upgrade?

**Comment by zacechola** [comment]
I think we need to see what's tripping people up often as the alphas are out there, but there are a few rules that need to be removed or rewritten given the new structures (containers can be nested now, panels are gone, etc. etc.).

We also need to have a bit of a meta discussion about maintaining multiple versions of Bootlint. As long as Bootstrap supports v3, we'll need to support it as well.

**Comment by kkirsche** [comment]
@zacechola do you think we want to run completely separate versions or provide a flag for each version and provide a single distribution of bootlint?

Certainly though some good things to rewrite and remove per the changes made in V4

**Comment by zacechola** [comment]
I'd vote for a flag. Getting into a meta discussion about structure here, but we could break out all the rules into their own files/directory structure and import them as needed. So we could have a big set of rules and then a v3 flag is just a set of rules and v4 is another set. That would also allow us to continue using the existing problem IDs. Something like `E004` would be in v3's set, but deprecated in v4's set.

cc @hnrch02

**Comment by hnrch02** [comment]
IIRC the goal for Bootlint v1 was full Bootstrap v4 support, don't know what we wanted to do about v3. Maybe @cvrebert can weigh in on that.

Supporting both v3 and v4 in the same Bootlint version would benefit users of both versions with new features and changes to the internals but would increase the maintenance burden for us. We could also maintain both 0.x and 1.x until Bootstrap v3 comes to its EOL and discontinue support for v3 in the 1.x series.

If we were to choose supporting both v3 and v4 in the same Bootlint version, I'd say the way to go would be to use some sort of auto-detection based on the classes used in the input. If the BS version could not be detected with certainty it could default to v4 and provide information to the user on how to switch to v3.

Anyway, I like the idea of splitting out each rule into its own file, like @zacechola said. I wanted to do that but didn't find the time yet.

**Comment by hnrch02** [comment]
Also, since Bootstrap v4 might still change quite a bit I'd argue that we should start working on Bootlint v1 when BS v4 approaches the first RC.

**Comment by chrismbarr** [comment]
Couldn't the version of bootstrap just be detected and then enable/disable different rule sets based on that?  Bootlint already warns about using bootstrap 4 with rule `W015`

**Comment by Herst** [comment]
Did someone already start working on this?

In any case, I made an attempt at some basic first steps towards support for it, see e.g. Herst/bootlint@7c64e98.

Again, those are only the very first steps.

**Comment by Herst** [comment]
Oops, somehow I based my changes on an old fork, I rebased it now: Herst/bootlint@10ed536

**Comment by Herst** [comment]
PR for tracking: #410

**Comment by nickmccurdy** [comment]
What's the current status of this? I heard some mentioning Bootstrap 4 alpha support, but is the beta unsupported? Can we add this information to the readme for transparency and to encourage people to contribute?

**Comment by Herst** [comment]
@nickmccurdy In my PR (see https://github.com/twbs/bootlint/pull/410 and <https://github.com/Herst/bootlint>) I added partial support for Bootstrap 4 at the state it was in when I worked on it (between beta 1 and 2). I would totally love for people to contribute.

**Comment by mackcoding** [comment]
Is there any update to bootlint for bs4? Really want to see this!

**Comment by Quix0r** [comment]
Please then also provide a 3.3.x-kompatible version on MaxCDN as not all will switch to BS4 (it is generally a good idea, of course, but off-topic here).

**Comment by rmullaney77** [comment]
Assumed the v4 compatible Bootlint would be available due to the age of this issue. Not to mention BS4 is no longer beta and actively working on 4.1.3.alpha 🤔

Any ETA on the 4.x compatible utility?

**Comment by Herst** [comment]
> Any ETA on the 4.x compatible utility?

No, at the moment there are mainly just some feature branches enabling compatibility for parts of it (e.g. the new grid): https://github.com/twbs/bootlint/branches

See also the pull requests and issues with **[v.next]** in the title. Help is always appreciated, e.g. by splitting up the wiki pages between v3 and v4+.

**Comment by stephenbe** [comment]
Hi @Herst,

Is there is any update on the ETA? Are we looking at 2020?

**Comment by Herst** [comment]
> Is there is any update on the ETA?

It's done when it's done. Help is always appreciated, even little stuff like updating the Wiki or such.

**Comment by nickmccurdy** [comment]
Could we have a flag to automatically enabled Bootstrap 4 rules and disable Bootstrap 3 only rules for now? I would be able to more easily test Bootstrap 4 support with that, but without it I mostly get false positives from Bootstrap 3 and it's difficult to isolate what I should be doing for Bootstrap 4.

**Comment by Herst** [comment]
This would basically disable all rules @nickmccurdy. The branch for bootstrap 4+ already exists, it just needs to be completed (which in some cases just means test cases and wiki pages). Any help is appreciated.

Look for the issues with v1 in the name.

**Comment by nickmccurdy** [comment]
Ah so they haven't been release yet? That's what I'm confused about.

**Comment by Herst** [comment]
See #431 to see its progress. Progress has been quite slow since I am the only person working on it and I haven't found the time (and sometimes motivation), other people in this project have done reviews.

**Comment by CRYPTO-GENIK** [comment]
Thanks for your work. I know it's thankless, but thank you. It helped me learn a lot over the years I've been using the bookmarklet.
